File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteRefresher.java
package com.mapbox.services.android.navigation.v5.navigation;

import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;

import java.util.Date;

class RouteRefresher {
  private final MapboxNavigation mapboxNavigation;
  private final RouteRefresh routeRefresh;
  private final long refreshIntervalInMilliseconds;
  private Date lastRefreshedDate;
  private boolean isChecking;
  private boolean isRefreshRouteEnabled;

  RouteRefresher(MapboxNavigation mapboxNavigation, RouteRefresh routeRefresh) {
    this.mapboxNavigation = mapboxNavigation;
    this.routeRefresh = routeRefresh;
    this.refreshIntervalInMilliseconds = mapboxNavigation.options().refreshIntervalInMilliseconds();
    this.lastRefreshedDate = new Date();
    this.isRefreshRouteEnabled = mapboxNavigation.options().enableRefreshRoute();
  }

  boolean check(Date currentDate) {
    if (isChecking || !isRefreshRouteEnabled) {
      return false;
    }
    long millisSinceLastRefresh = currentDate.getTime() - lastRefreshedDate.getTime();
    return millisSinceLastRefresh > refreshIntervalInMilliseconds;
  }

  void refresh(RouteProgress routeProgress) {
    updateIsChecking(true);
    routeRefresh.refresh(routeProgress, new RouteRefresherCallback(mapboxNavigation, this));
  }

  void updateLastRefresh(Date date) {
    lastRefreshedDate = date;
  }

  void updateIsChecking(boolean isChecking) {
    this.isChecking = isChecking;
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/RouteRefresher.kt
package com.mapbox.services.android.navigation.v5.internal.navigation

import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation
import com.mapbox.services.android.navigation.v5.navigation.RouteRefresh
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
import java.util.Date

internal class RouteRefresher(
    private val mapboxNavigation: MapboxNavigation,
    private val routeRefresh: RouteRefresh
) : RouteRefresherInterface {
    private val refreshIntervalInMilliseconds: Long = mapboxNavigation.options().refreshIntervalInMilliseconds()
    private val isRefreshRouteEnabled: Boolean = mapboxNavigation.options().enableRefreshRoute()
    private var lastRefreshedDate = Date()
    private var isChecking: Boolean = false

    override fun check(currentDate: Date): Boolean {
        if (isChecking || !isRefreshRouteEnabled) {
            return false
        }
        val millisSinceLastRefresh = currentDate.time - lastRefreshedDate.time
        return millisSinceLastRefresh > refreshIntervalInMilliseconds
    }

    override fun refresh(routeProgress: RouteProgress) {
        updateIsChecking(true)
        routeRefresh.refresh(routeProgress, RouteRefresherCallback(mapboxNavigation, this))
    }

    override fun updateLastRefresh(date: Date) {
        lastRefreshedDate = date
    }

    override fun updateIsChecking(isChecking: Boolean) {
        this.isChecking = isChecking
    }
}
